salt/transport/ipc.py

```diff
@@ -0,0 +1,333 @@
+# -*- coding: utf-8 -*-
+'''
+IPC transport classes
+'''
+
+# Import Python libs
+from __future__ import absolute_import
+import logging
+import socket
+import msgpack
+import weakref
+
+# Import Tornado libs
+import tornado
+import tornado.gen
+import tornado.netutil
+from tornado.ioloop import IOLoop
+from tornado.iostream import IOStream
+
+# Import Salt libs
+import salt.transport.client
+import salt.transport.frame
+
+log = logging.getLogger(__name__)
+
+
+class IPCServer(object):
+    '''
+    A Tornado IPC server very similar to Tornado's TCPServer class
+    but using either UNIX domain sockets or TCP sockets
+    '''
+    def __init__(self, opts, io_loop=None, stream_handler=None):
+        '''
+        Create a new Tornado IPC server
+
+        :param IOLoop io_loop: A Tornado ioloop to handle scheduling
+        :param func stream_handler: A function to customize handling of an
+                                    incoming stream.
+        '''
+        self.opts = opts
+        self._started = False
+        self.stream_handler = stream_handler
+
+        # Placeholders for attributes to be populated by method calls
+        self.stream = None
+        self.sock = None
+        self.io_loop = None
+
+    def start(self, socket_path):
+        '''
+        Perform the work necessary to start up a Tornado IPC server
+
+        Blocks until socket is established
+
+        :param str socket_path: Path on the filesystem for the socket to bind to.
+                                This socket does not need to exist prior to calling
+                                this method, but parent directories should.
+        '''
+        # Start up the ioloop
+        log.trace('IPCServer: binding to socket: {0}'.format(socket_path))
+        self.sock = tornado.netutil.bind_unix_socket(socket_path)
+
+        if self.io_loop is None:
+            self.io_loop = IOLoop.current()
+
+        tornado.netutil.add_accept_handler(
+            self.sock,
+            self.handle_connection,
+            io_loop=self.io_loop,
+        )
+        self._started = True
+
+    @tornado.gen.coroutine
+    def handle_stream(self, stream):
+        '''
+        Override this to handle the streams as they arrive
+
+        :param IOStream stream: An IOStream for processing
+
+        See http://tornado.readthedocs.org/en/latest/iostream.html#tornado.iostream.IOStream
+        for additional details.
+        '''
+        while True:
```

Comment:
We probably want some try/except wrapped in a close. If the client sends us garbage we should log an error and close their stream. Right now we'll stacktrace :)